### PR TITLE
Work around the failures at updating Git deployment tags.

### PR DESCRIPTION
**Pull request by artsyjian**
The "post-deploy" step that updates Git tags, has been failing once in a while (at git fetch tags). Its failure causes`Deployment.update()` to raise `HokusaiError`, which causes CircleCI to report failure for the entire deploy. Here's an example:

https://app.circleci.com/pipelines/github/artsy/gravity/3684/workflows/965fc470-8e34-47db-a518-55b3bea59e02/jobs/11114

From discussion we conclude that the step is not critical to deploys. It's good to have the Git tags, but failing to create them should not fail the deploy. Therefore:

- Prevent `HokusaiError` from getting triggered by this step's failure. This is achieved by not setting `post_deploy_success` to `False` when the step fails.
- When catching `CalledProcessError` exception, print its `output` attribute which might show the actual Git error.
- Retry the step up to 3 times at 3-second intervals, as best effort to have this step succeed.

### CODE REVIEW

**Comment by artsyjian** [comment]
When the step fails, the output will look like this:

`(hokusai)$ hokusai pipeline promote --update-config                                                                                                                                                                                               [45/967]
Deploying sha256:9687df043af01586c74b5e8105596bf2739533cdcb795564374df2cb6b4f7a9f to production...

Running pre-deploy hook 'echo Ohai'...

Ohai
Patching Deployments in spec /home/jian/coderepos/hokusai-sandbox/hokusai/production.yml.j2 with image digest sha256:9687df043af01586c74b5e8105596bf2739533cdcb795564374df2cb6b4f7a9f

Applying patched spec /home/jian/coderepos/hokusai-sandbox/.hokusai-tmp/tmpMNoZ_b...

deployment.apps "hokusai-sandbox-web" configured
service "hokusai-sandbox-web" unchanged
Waiting for deployment rollouts to complete...
Waiting for rollout to finish: 1 old replicas are pending termination...
Waiting for rollout to finish: 1 old replicas are pending termination...
deployment "hokusai-sandbox-web" successfully rolled out
Running post-deploy hook 'echo Bai'...

Bai
Updating ECR deployment tags in 585031190124.dkr.ecr.us-east-1.amazonaws.com/hokusai-sandbox...

Updated ECR tag staging -> production
Updated ECR tag staging -> production--2020-05-31--16-38-05

Creating Git deployment tags 'production', 'production--2020-05-31--16-38-05', and push them to foo...
Attempt# 1.

WARNING: Creating Git deployment tags failed due to the error: 'Command 'git fetch foo --tags' returned non-zero exit status 128'


WARNING: Additional error information if any: 'fatal: 'foo' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
'

Creating Git deployment tags 'production', 'production--2020-05-31--16-38-05', and push them to foo...
Attempt# 2.

WARNING: Creating Git deployment tags failed due to the error: 'Command 'git fetch foo --tags' returned non-zero exit status 128'


WARNING: Additional error information if any: 'fatal: 'foo' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
'

Creating Git deployment tags 'production', 'production--2020-05-31--16-38-05', and push them to foo...
Attempt# 3.

WARNING: Creating Git deployment tags failed due to the error: 'Command 'git fetch foo --tags' returned non-zero exit status 128'


WARNING: Additional error information if any: 'fatal: 'foo' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
'

Failed all attempts at pushing Git deployment tags! Please do it manually.

Deployment succeeded!
Promoted staging to production
`

**Comment by artsyjian** [comment]
> This looks like a great improvement. I wonder if there is a corresponding test file that can cover this?

There appears to be no existing test that covers `Deployment.update()`. I will run it by @izakp tomorrow.

File: hokusai/services/deployment.py [hunk 1]
```diff
@@ -164,19 +165,31 @@ def update(self, tag, constraint, git_remote, timeout, update_config=False, file
...
-          print_yellow("The tag %s has been rolled out.  However, you should create the Git tags '%s' and '%s' manually, or re-run this deployment." % (tag, deployment_tag, self.context), newline_after=True)
-          post_deploy_success = False
+        # Update git tags. Try up to 3 times, at 3 second intervals. Failure does not fail deployment.
+        git_tag_sucess = False
+        attempts = 0
+        while ((not git_tag_sucess) and (attempts < 3)):
+          try:
+            attempts += 1
+            print_green("Creating Git deployment tags '%s', '%s', and push them to %s..." % (self.context, deployment_tag, remote))
+            print_green("Attempt# %s." %attempts)
```
Comment (izakp) [inline hokusai/services/deployment.py:175]: `s/%attempts/% attempts`

File: hokusai/services/deployment.py [hunk 2]
```diff
@@ -164,19 +165,31 @@ def update(self, tag, constraint, git_remote, timeout, update_config=False, file
...
+            shout("git fetch %s --tags" % remote)
+            shout("git tag -f %s %s" % (self.context, tag), print_output=True)
+            shout("git tag -f %s %s" % (deployment_tag, tag), print_output=True)
+            shout("git push -f --no-verify %s refs/tags/%s" % (remote, self.context), print_output=True)
+            print_green("Updated Git tag %s -> %s" % (tag, self.context))
+            shout("git push -f --no-verify %s refs/tags/%s" % (remote, deployment_tag), print_output=True)
+            print_green("Updated Git tag %s -> %s" % (tag, deployment_tag), newline_after=True)
+          except CalledProcessError as e:
+            print_yellow("WARNING: Creating Git deployment tags failed due to the error: '%s'" % str(e), newline_before=True, newline_after=True)
```
Comment (izakp) [inline hokusai/services/deployment.py:184]: 👍 on logging the `output` attribute for the CalledProcessError raised.  I think we can drop this line that logs `str(e)`

File: hokusai/services/deployment.py [hunk 3]
```diff
@@ -164,19 +165,31 @@ def update(self, tag, constraint, git_remote, timeout, update_config=False, file
...
+            # If subprocess.check_output() was called, the actual error is in CalledProcessError's 'output' attribute.
+            print_yellow("WARNING: Additional error information if any: '%s'" % e.output, newline_before=True, newline_after=True)
+            time.sleep(3)
+          else:
+            git_tag_sucess = True
```
Comment (izakp) [inline hokusai/services/deployment.py:189]: #minor In favor of moving `git_tag_sucess = True` to line 183.  I find `try ... except ... else` statements less clear in terms of control flow than simply `try ... except`

File: hokusai/services/deployment.py [hunk 4]
```diff
@@ -164,19 +165,31 @@ def update(self, tag, constraint, git_remote, timeout, update_config=False, file
...
-          print_yellow("WARNING: Creating Git deployment tags failed due to the error: '%s'" % str(e), newline_before=True, newline_after=True)
-          print_yellow("The tag %s has been rolled out.  However, you should create the Git tags '%s' and '%s' manually, or re-run this deployment." % (tag, deployment_tag, self.context), newline_after=True)
-          post_deploy_success = False
+        # Update git tags. Try up to 3 times, at 3 second intervals. Failure does not fail deployment.
+        git_tag_sucess = False
+        attempts = 0
+        while ((not git_tag_sucess) and (attempts < 3)):
+          try:
+            attempts += 1
+            print_green("Creating Git deployment tags '%s', '%s', and push them to %s..." % (self.context, deployment_tag, remote))
```
Comment (izakp) [inline hokusai/services/deployment.py:174]: #minor `s/push/pushing`

File: hokusai/VERSION [hunk 5]
```diff
@@ -1 +1 @@
-0.5.9
+0.5.10
```
Comment (izakp) [inline hokusai/VERSION:1]: Would not update the version yet.  The normal workflow is to create a "prepare version x.x.x" commit / PR including updates to the Changelog once we are ready to release - see dfea0c7545486d52c9a08f27915d64bc7165631b

File: hokusai/services/deployment.py (hunk 1 above)
Comment (artsyjian) [reply hokusai/services/deployment.py:175]: done.

**Comment by izakp** [comment]
> There appears to be no existing test that covers Deployment.update(). I will run it by @izakp tomorrow.

No there is no integration test for deployment rollout - basic kubernetes CRUD operations are tested in https://github.com/artsy/hokusai/blob/master/test/integration/test_kubernetes.py but because this change is changing only the post-deploy git tagging logic, I think it would be better unit tested, at least to the level of mocking calls to `git`.

I think we should eventually refactor `Deployment.update()` and break it into other `Deployment` methods such as `Deployment._pre_deploy()` / `Deployment._patch()` / `Deployment._rollout()` / `Deployment._post_deploy()` / `Deployment._push_image_tags()` / `Deployment._push_git_tags()` so perhaps breaking this out to `Deployment._push_git_tags()` is a good start?

File: hokusai/services/deployment.py (hunk 2 above)
Comment (artsyjian) [reply hokusai/services/deployment.py:184]: done.

File: hokusai/services/deployment.py (hunk 4 above)
Comment (artsyjian) [reply hokusai/services/deployment.py:174]: done.

File: hokusai/services/deployment.py (hunk 3 above)
Comment (artsyjian) [reply hokusai/services/deployment.py:189]: done.

File: hokusai/VERSION (hunk 5 above)
Comment (artsyjian) [reply hokusai/VERSION]: Reverted VERSION.

**Comment by artsyjian** [comment]
Cut ticket on suggestion to break Deployment.update() into parts, plus adding unit test for each part.
https://artsyproduct.atlassian.net/browse/PLATFORM-2416

**Comment by artsyjian** [comment]
Thanks @eessex, @izakp, for the great reviews. I cut a ticket for refactoring/uni-testing Deployment.update(). The other comments are resolved. How do we proceed? The change addresses a chronic issue which we are eager to put a stop to. But such a small change does not seem worthy of a version bump (0.5.9 -> 0.5.10). Do we wait, collect more changes, and release them all at once? Or, do we roll this out to a beta version and start using it internally?

**Comment by izakp** [comment]
I am for getting this merged to master so we can start using a beta version.  I think this fix and https://github.com/artsy/hokusai/pull/205 are enough for a patch version bump.

**Pull request merged**